hydro_lang/src/viz/json.rs

```diff
@@ -0,0 +1,1119 @@
+        // Remove everything before the last "::" to get just the function name
+        fn_name.split("::").last().unwrap_or(fn_name).to_string()
+    }
+}
+
+impl<W> HydroGraphWrite for HydroJson<W>
+where
+    W: Write,
+{
+    type Err = super::render::GraphWriteError;
+
+    fn write_prologue(&mut self) -> Result<(), Self::Err> {
+        // Clear any existing data
+        self.nodes.clear();
+        self.edges.clear();
+        self.locations.clear();
+        self.node_locations.clear();
+        self.edge_count = 0;
+        Ok(())
+    }
+
+    fn write_node_definition(
+        &mut self,
+        node_id: usize,
+        node_label: &super::render::NodeLabel,
+        node_type: HydroNodeType,
+        location_id: Option<usize>,
+        location_type: Option<&str>,
+        backtrace: Option<&Backtrace>,
+    ) -> Result<(), Self::Err> {
+        // Create the full label string using DebugExpr::Display for expressions
+        let full_label = match node_label {
+            super::render::NodeLabel::Static(s) => s.clone(),
+            super::render::NodeLabel::WithExprs { op_name, exprs } => {
+                if exprs.is_empty() {
+                    format!("{}()", op_name)
+                } else {
+                    // This is where DebugExpr::Display gets called with q! macro cleanup
+                    let expr_strs: Vec<String> = exprs.iter().map(|e| e.to_string()).collect();
+                    format!("{}({})", op_name, expr_strs.join(", "))
+                }
+            }
+        };
+
+        // Always extract short label for UI toggle functionality
+        let short_label = super::render::extract_short_label(&full_label);
+
+        // If short and full labels are the same or very similar, enhance the full label
+        // Use saturating comparison to avoid underflow when full_label is very short
+        let full_len = full_label.len();
+        let enhanced_full_label = if short_label.len() >= full_len.saturating_sub(2) {
+            // If they're nearly the same length, add more context to full label
+            match short_label.as_str() {
+                "inspect" => "inspect [debug output]".to_string(),
+                "persist" => "persist [state storage]".to_string(),
+                "tee" => "tee [branch dataflow]".to_string(),
+                "delta" => "delta [change detection]".to_string(),
+                "spin" => "spin [delay/buffer]".to_string(),
+                "send_bincode" => "send_bincode [send data to process/cluster]".to_string(),
+                "broadcast_bincode" => {
+                    "broadcast_bincode [send data to all cluster members]".to_string()
+                }
+                "source_iter" => "source_iter [iterate over collection]".to_string(),
+                "source_stream" => "source_stream [receive external data stream]".to_string(),
+                "network(recv)" => "network(recv) [receive from network]".to_string(),
```

Comment:
We should really only be rendering a single node for network rather than splitting halves. But okay to defer.